packages/grafana-ui/src/components/uPlot/plugins/KeyboardPlugin.tsx

```diff
@@ -0,0 +1,163 @@
+import React, { useLayoutEffect } from 'react';
+import uPlot from 'uplot';
+import { UPlotConfigBuilder } from '../config/UPlotConfigBuilder';
+
+interface KeyboardPluginProps {
+  config: UPlotConfigBuilder; // onkeypress, onkeyup, onkeydown (triggered by vizlayout handlers)
+}
+
+const PIXELS_PER_MS = 0.1 as const;
+const SHIFT_MULTIPLIER = 2 as const;
+const KNOWN_KEYS = new Set(['ArrowRight', 'ArrowLeft', 'ArrowUp', 'ArrowDown', 'Shift', ' ']);
+
+function minMax(n: number, min: number, max: number) {
+  return Math.min(Math.max(min, n), max);
+}
+
+const initHook = (u: uPlot) => {
+  let vizLayoutViz: HTMLElement | null = u.root.closest('[tabindex]');
+  let pressedKeys = new Set<string>();
+  let dragStartX: number | null = null;
+  let keysLastHandledAt: number | null = null;
+
+  if (!vizLayoutViz) {
+    return;
+  }
+
+  const moveCursor = (dx: number, dy: number) => {
+    const { cursor } = u;
+    if (cursor.left === undefined) {
```

Comment:
**Question:** should we check for `cursor.top` here as well, or we are just checking if the object is not empty?